src/components/select/select.js

```diff
@@ -0,0 +1,304 @@
+// @flow
+import * as React from 'react';
+import {findDOMNode} from 'react-dom';
+import {
+  Root as StyledRoot,
+  Input as StyledInput,
+  InputContainer as StyledInputContainer,
+  Tag as StyledTag,
+  SearchIcon as StyledSearchIcon,
+} from './styled-components';
+
+import {Input as InputComponent} from '../input';
+import {ICON, TYPE} from './constants';
+import SelectDropDown from './dropdown';
+import type {DefaultPropsT, OptionT, PropsT, StatelessStateT} from './types';
+import {getOverride} from '../../helpers/overrides';
+
+class Select extends React.Component<PropsT, StatelessStateT> {
+  static defaultProps: DefaultPropsT = {
+    overrides: {},
+    selectedOptions: [],
+    onSelect: () => {},
+    onUnSelect: () => {},
+    onClearAll: () => {},
+    onKeyUp: () => {},
+    onBlur: () => {},
+    onFocus: () => {},
+    onMouseEnter: () => {},
+    onMouseLeave: () => {},
+    onMouseDown: () => {},
+    onMouseUp: () => {},
+    error: '',
+    autoFocus: false,
+  };
+
+  state = {
+    textValue: '',
+    selectedOptions: [],
+    isDropDownOpen: this.props.type === TYPE.SEARCH,
+    ...this.props,
+  };
+
+  constructor(props: PropsT) {
+    super(props);
+    if (props.type === TYPE.SELECT) {
+      const {selectedOptions} = this.state;
+      const {options = []} = props;
+      selectedOptions.forEach(selectedOption => {
+        // selected option is not in all options and needs to be added on top
+        if (!options.find(selected => selected.id === selectedOption.id)) {
+          options.unshift({
+            id: selectedOption.id,
+            label: this.getOptionLabel(selectedOption),
+          });
+        }
+      });
+    }
+  }
+
+  componentDidMount() {
+    const document = require('global/document');
+    document.addEventListener('click', this.handleClickEvent, {
+      capture: true,
+    });
+  }
+
+  componentWillUnmount() {
+    const document = require('global/document');
```

Comment:
Can we move this require to the top of the file?

Reply:
yes, seems like it's friendly to server-side rendering

Reply:
Seems like these still haven't been moved?

Reply:
I took it back to inline as @rtsao said I should avoid import for those, which work only on client-side

Reply:
Unless you guys had a separate conversation, it sounds like @rtsao said to just use eslint comments:

```
componentWillUnmount() {
  /* eslint-disable-next-line noundef  */
  document.addEventListener('click', this.handleClickEvent, {
      capture: true,
    });
}
```

Personally I'd be fine with just adding a `/* global document */` to the top of the file, and relying on tests to catch cases where we're incorrectly using document server-side.

Reply:
Yeah, I think including this as an actual dependency is pure bloat